pinot-plugins/pinot-file-system/pinot-adls/src/main/java/org/apache/pinot/plugin/filesystem/AzureGen2PinotFS.java

```diff
@@ -0,0 +1,447 @@
+    try {
+      PathProperties pathProperties = getPathProperties(uri);
+      OffsetDateTime offsetDateTime = pathProperties.getLastModified();
+      Timestamp timestamp = Timestamp.valueOf(offsetDateTime.atZoneSameInstant(ZoneOffset.UTC).toLocalDateTime());
+      return timestamp.getTime();
+    } catch (DataLakeStorageException e) {
+      throw new IOException("Failed while checking lastModified time for : " + uri, e);
+    }
+  }
+
+  @Override
+  public boolean touch(URI uri) throws IOException {
+    // The following data lake gen2 API provides a way to update file properties including last modified time.
+    // https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update
+    // However, action = "setProperties" is available in REST API but not available in Java SDK yet.
+    //
+    // For now, directly use Blob service's API to get the same effect.
+    // https://docs.microsoft.com/en-us/rest/api/storageservices/set-file-properties
+    try {
+      DataLakeFileClient fileClient = _fileSystemClient.getFileClient(convertUriToAzureStylePath(uri));
+      PathProperties pathProperties = fileClient.getProperties();
+      fileClient.setHttpHeaders(getPathHttpHeaders(pathProperties));
+      return true;
+    } catch (DataLakeStorageException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public InputStream open(URI uri) throws IOException {
+    // Use Blob API since read() function from Data Lake Client currently takes "OutputStream" as an input and
+    // flush bytes to an output stream. This needs to be piped back into input stream to implement this function.
+    // On the other hand, Blob API directly allow you to open the input stream.
+    BlobClient blobClient = _blobServiceClient.getBlobContainerClient(_fileSystemClient.getFileSystemName())
+        .getBlobClient(convertUriToAzureStylePath(uri));
+
+    return blobClient.openInputStream();
+    // Another approach is to download the file to the local disk to a temp path and return the file input stream. In
+    // this case, we need to override "close()" and delete temp file.
+  }
+
+  private boolean copySrcToDst(URI srcUri, URI dstUri) throws IOException {
+    try (InputStream inputStream = open(srcUri)) {
+      return copyInputStreamToDst(inputStream, dstUri);
+    }
+  }
+
+  /**
+   * Helper function to copy input stream to destination URI.
+   *
+   * NOTE: the caller has to close the input stream.
+   *
+   * @param inputStream input stream that will be written to dstUri
+   * @param dstUri destination URI
+   * @return true if the copy succeeds
+   */
+  private boolean copyInputStreamToDst(InputStream inputStream, URI dstUri) throws IOException {
+    int bytesRead;
+    long totalBytesRead = 0;
+    byte[] buffer = new byte[BUFFER_SIZE];
+    DataLakeFileClient fileClient = _fileSystemClient.createFile(convertUriToAzureStylePath(dstUri));
+    try {
+      MessageDigest md5File = MessageDigest.getInstance("MD5");
```

Comment:
Seems like this part of the code can be refactored, since there is one more place doing something similar?

Reply:
discussed offline. resolving this.